src/machines/java/com/enderio/machines/common/blockentity/base/VacuumMachineEntity.java

```diff
@@ -0,0 +1,153 @@
+package com.enderio.machines.common.blockentity.base;
+
+import com.enderio.api.io.IIOConfig;
+import com.enderio.api.io.IOMode;
+import com.enderio.base.common.particle.RangeParticleData;
+import com.enderio.base.common.util.AttractionUtil;
+import com.enderio.core.common.sync.BooleanDataSlot;
+import com.enderio.core.common.sync.IntegerDataSlot;
+import com.enderio.core.common.sync.SyncMode;
+import com.enderio.machines.common.io.FixedIOConfig;
+import net.minecraft.core.BlockPos;
+import net.minecraft.server.level.ServerLevel;
+import net.minecraft.server.level.ServerPlayer;
+import net.minecraft.world.entity.Entity;
+import net.minecraft.world.level.Level;
+import net.minecraft.world.level.block.entity.BlockEntityType;
+import net.minecraft.world.level.block.state.BlockState;
+import net.minecraft.world.phys.AABB;
+import net.minecraft.world.phys.Vec3;
+
+import java.lang.ref.WeakReference;
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.function.Predicate;
+
+public abstract class VacuumMachineEntity<T extends Entity> extends MachineBlockEntity {
+    private static final double COLLISION_DISTANCE_SQ = 1 * 1;
+    protected static final double SPEED = 0.025;
+    protected static final double SPEED_4 = SPEED * 4;
+    private static final int MAX_RANGE = 6;
+    private int range = 6;
+    protected float rCol = 1;
+    protected float gCol = 0;
+    protected float bCol = 0;
+
+    private boolean rangeVisible = false;
+    private List<WeakReference<T>> entities = new ArrayList<>();
+    private Class<T> clazz;
+
+    public VacuumMachineEntity(BlockEntityType<?> pType, BlockPos pWorldPosition, BlockState pBlockState, Class<T> clazz) {
+        super(pType, pWorldPosition, pBlockState);
+        this.clazz = clazz;
+        add2WayDataSlot(new IntegerDataSlot(this::getRange, this::setRange, SyncMode.GUI));
+        add2WayDataSlot(new BooleanDataSlot(this::isShowingRange, this::shouldShowRange, SyncMode.GUI));
+    }
+
+    @Override
+    public void serverTick() {
+        if (this.getRedstoneControl().isActive(level.hasNeighborSignal(worldPosition))) {
+            this.attractEntities(this.getLevel(), this.getBlockPos(), this.range);
+        }
+
+        if (this.isShowingRange()) {
+            generateParticle(new RangeParticleData(getRange(), this.rCol, this.gCol, this.bCol),
+                new Vec3(getBlockPos().getX(), getBlockPos().getY(), getBlockPos().getZ()));
+        }
+        super.serverTick();
+    }
+
+    @Override
+    public void clientTick() {
+        if (this.getRedstoneControl().isActive(level.hasNeighborSignal(worldPosition))) {
+            this.attractEntities(this.getLevel(), this.getBlockPos(), this.range);
+        }
+        super.clientTick();
+    }
+
+    @Override
+    protected IIOConfig createIOConfig() {
+        return new FixedIOConfig(IOMode.PUSH);
```

Comment:
In 1.12, Both the vacuum block don't have an option to configure automatic push/pull. However, XP vacuum has a very small internal buffer and thus requires an external storage connected to it. It pushes automatically to a tank next to it. On the other hand, Vacuum chest doesn't push anything automatically.

Reply:
I felt like giving these blocks also control over the sides be a good idea. And as a correction, from what I could see in the 1.12 code, the buffer of the xp vacuum is really large (to the point it may need a nerf).

Reply:
[This](https://github.com/SleepyTrousers/EnderIO/blob/9249b7e01fd73c7bc92a2fd60e768c1ef5b39fd3/enderio-machines/src/main/java/crazypants/enderio/machines/machine/vacuum/xp/TileXPVacuum.java#L52) says that 1.12 had same capacity

Reply:
Should I keep it default or change it ?